Enable building jna for Darwin arm64

fkistner opened the pull request:
We are interested in getting JNA running on Darwin arm64.

I was able to get JNA compiling and working for the use cases most important to us with these changes and
`env JDK_HOME=… SDKROOT=… ant -DCC=clang -Dmake.OPTS="DYNAMIC_LIBFFI=true DARWIN_USE_SDK_LIBFFI=true"`.
Nevertheless, I am not too familiar with JNA and would love to hear your feedback and whether you are interesting in the changes in their current state.

Summary of the changes:
- Introduced checks in the Makefile to determine which architectures should be built on Darwin and added flag to allow linking against the system `libffi`.
- Big Sur uses a shared dylib cache to lookup system libraries and frameworks. Checking existence in the file system will fail, but loading them will succeed nevertheless.
   Adapted `LibraryLoadTest` accordingly.
- Arm64 calling convention requires varargs to always be passed on the stack. JNA's and ffi's argument handling logic seems to handle this case correctly, but it is important that the Java definitions match their native counterparts exactly.
   Adapted `VarArgsTest` accordingly.

<details>
  <summary>Unfortunately, some tests still fail…</summary>

```
Testsuite: com.sun.jna.AnnotatedLibraryTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1,818 sec

Testsuite: com.sun.jna.ArgumentsMarshalNullableTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,303 sec

Testsuite: com.sun.jna.ArgumentsMarshalTest
Tests run: 41, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,432 sec

Testsuite: com.sun.jna.ArgumentsWrappersMarshalTest
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,293 sec

Testsuite: com.sun.jna.BufferArgumentsMarshalTest
Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,369 sec

Testsuite: com.sun.jna.ByReferenceArgumentsTest
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,296 sec

Testsuite: com.sun.jna.ByReferenceToStringTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,262 sec

Testsuite: com.sun.jna.CallbacksTest
Exception in thread "Thread-0" JNA: error while handling callback exception, continuing
Exception in thread "main" JNA: error while handling callback exception, continuing
Tests run: 53, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9,849 sec

------------- Standard Error -----------------
java.lang.StackOverflowError
Warning: JVM did not GC Thread mapping after native thread terminated
------------- ---------------- ---------------
Testsuite: com.sun.jna.DefaultMethodInvocationTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,341 sec

Testsuite: com.sun.jna.DirectArgumentsMarshalNullableTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,315 sec

Testsuite: com.sun.jna.DirectArgumentsMarshalTest
Tests run: 41, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,384 sec

Testsuite: com.sun.jna.DirectArgumentsWrappersMarshalTest
Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,395 sec

Testsuite: com.sun.jna.DirectBufferArgumentsMarshalTest
Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,379 sec

Testsuite: com.sun.jna.DirectByReferenceArgumentsTest
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,329 sec

Testsuite: com.sun.jna.DirectCallbacksTest
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (os_bsd_zero.cpp:241), pid=75646, tid=5635
#  fatal error: caught unhandled signal 11 at address 0x0400000001000009
#
# JRE version: OpenJDK Runtime Environment (14.0.1) (build 14.0.1-internal+0)
# Java VM: OpenJDK 64-Bit Zero VM (14.0.1-internal+0, interpreted mode, serial gc, bsd-aarch64)
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# hs_err_pid75646.log
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
Testsuite: com.sun.jna.DirectCallbacksTest
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0 sec

Testcase: com.sun.jna.DirectCallbacksTest:testUnionByValueCallbackArgument:	Caused an ERROR
Forked Java VM exited abnormally. Please note the time in the report does not reflect the time until the VM exit.
junit.framework.AssertionFailedError: Forked Java VM exited abnormally. Please note the time in the report does not reflect the time until the VM exit.
	at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)


Test com.sun.jna.DirectCallbacksTest FAILED (crashed)
Testsuite: com.sun.jna.DirectReturnTypesTest
Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,337 sec

Testsuite: com.sun.jna.DirectStructureByValueTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,317 sec

Testsuite: com.sun.jna.DirectTest
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,335 sec

Testsuite: com.sun.jna.DirectTypeMapperTest
Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,332 sec

Testsuite: com.sun.jna.ELFAnalyserTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,199 sec

Testsuite: com.sun.jna.FunctionTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,283 sec

Testsuite: com.sun.jna.HeadlessLoadLibraryTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,265 sec

Testsuite: com.sun.jna.IntegerTypeTest
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,288 sec

Testsuite: com.sun.jna.JNALoadTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,58 sec

Testsuite: com.sun.jna.LastErrorTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,345 sec

Testsuite: com.sun.jna.LibraryLoadTest
Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,787 sec

Testsuite: com.sun.jna.MemoryTest
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 13,652 sec

Testsuite: com.sun.jna.NativeLibraryTest
Tests run: 22, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2,393 sec

Testsuite: com.sun.jna.NativeTest
Tests run: 33, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2,162 sec

------------- Standard Error -----------------
Jul 21, 2020 3:07:10 PM com.sun.jna.Native getCharset
WARNING: JNA Warning: Encoding 'unsupported' is unsupported (unsupported)
Jul 21, 2020 3:07:11 PM com.sun.jna.Native getCharset
WARNING: JNA Warning: Using fallback encoding UTF-8
------------- ---------------- ---------------
Testsuite: com.sun.jna.NativedMappedTest
Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,303 sec

Testsuite: com.sun.jna.PerformanceTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,153 sec

Testsuite: com.sun.jna.PlatformTest
Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,164 sec

Testsuite: com.sun.jna.PointerBufferTest
Tests run: 16, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,255 sec

Testsuite: com.sun.jna.PointerTest
Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,273 sec

Testsuite: com.sun.jna.PrematureGCTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,623 sec

Testsuite: com.sun.jna.ReturnTypesTest
Tests run: 19, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,338 sec

Testsuite: com.sun.jna.StructureBufferFieldTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,276 sec

Testsuite: com.sun.jna.StructureByValueTest
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,307 sec

Testsuite: com.sun.jna.StructureFieldOrderInspectorTest
Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,986 sec

Testsuite: com.sun.jna.StructureTest
Tests run: 90, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,578 sec

Testsuite: com.sun.jna.TypeMapperTest
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,314 sec

Testsuite: com.sun.jna.UnionTest
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,327 sec

Testsuite: com.sun.jna.VMCrashProtectionTest
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,278 sec

Testsuite: com.sun.jna.VarArgsCheckerTest
Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0,159 sec

Testsuite: com.sun.jna.VarArgsTest
Tests run: 9, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0,309 sec

Testcase: testShortVarArgs(com.sun.jna.VarArgsTest):	FAILED
16-bit integer varargs not added correctly expected:<3> but was:<-697827325>
junit.framework.AssertionFailedError: 16-bit integer varargs not added correctly expected:<3> but was:<-697827325>
	at com.sun.jna.VarArgsTest.testShortVarArgs(VarArgsTest.java:66)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)


Test com.sun.jna.VarArgsTest FAILED
```

</details>

matthiasblaesing commented on native/Makefile:
```diff
@@ -373,19 +383,26 @@ endif
 
 ifeq ($(OS),darwin)
-XCODE_VERSION=$(shell xcodebuild -version | grep Xcode | sed 's/^Xcode \([1-9]\).*/\1/g')
+XCODE_VERSION=$(shell xcodebuild -version | grep Xcode | sed 's/^Xcode \([1-9][0-9]*\).*/\1/g')
```
Xcode version became two digit?

matthiasblaesing commented on native/Makefile:
```diff
@@ -373,19 +383,26 @@ endif
 
 ifeq ($(OS),darwin)
-XCODE_VERSION=$(shell xcodebuild -version | grep Xcode | sed 's/^Xcode \([1-9]\).*/\1/g')
+XCODE_VERSION=$(shell xcodebuild -version | grep Xcode | sed 's/^Xcode \([1-9][0-9]*\).*/\1/g')
```
Xcode became two digit?

fkistner replied:
Yes, the newest version of Xcode is 12:
https://developer.apple.com/documentation/xcode-release-notes/xcode-12-beta-release-notes

matthiasblaesing commented on native/Makefile:
```diff
@@ -373,19 +383,26 @@ endif
 FFI_ENV += CC="$(CC)" CFLAGS="-arch $(ARCH) $(ISYSROOT) $(COPT) $(CDEBUG)" CPPFLAGS="$(CDEFINES)" LD="$(LD) -arch $(ARCH)"
 ALT_ARCHS=
 ifneq ($(ARCH),i386)
+ifeq ($(shell echo "int main(void) { return 0; }" | $(CC) $(LOC_CC_OPTS) -arch i386 $(CFLAGS) -x c - -o /dev/null &>/dev/null; echo $$?),0)
```
Could you explain what this should do?

fkistner replied:
This tries to compile a tiny binary for the given architecture. If the architecture is not supported, the check will fail.

fkistner replied:
Reverted this. Shouldn't be necessary to run compiler. The deciding factor seems to be Xcode 10.

matthiasblaesing commented on native/Makefile:
```diff
@@ -401,7 +418,13 @@ JNISFX=.jnilib
+endif
+ifeq ($(shell echo "int main(void) { return 0; }" | $(CC) $(LOC_CC_OPTS) -arch arm64 $(CFLAGS) -x c - -o /dev/null &>/dev/null; echo $$?),0)
+ARCHFLAGS+=-arch arm64
+endif
```
This looks as if i386 and/or arm64 are conditional - is this correct? Is my interpretation correct, that this basicly says: "If the C compiler can compile for 32bit, then do it, if it can compile for arm64, then do it. Is this correct?

fkistner replied:
Yes, correct.

matthiasblaesing replied:
Currently we build fatbinaries with x86 + x86-64. Is there an XCode version out there, that is able to create a tipple binary: x86 + x86_64 + arm64? Or would now be a good time to split the binaries for mac OS as all other architectures? This should be taken to the mailing list, as it will be incompatible/needs some deeper thoughts.

matthiasblaesing commented on native/Makefile:
```diff
@@ -410,16 +433,18 @@ endif
 # JAWT linkage handled by -framework JavaVM
-LIBS=
+LIBS+=-framework JavaVM
+endif
```
Is linking against JavaVM not required anymore?

fkistner replied:
Apple does not ship a `JavaVM` framework anymore.

matthiasblaesing replied:
So do we need the JavaVM framework? Or could we drop it from the build entirely?

fkistner replied:
It seems to me this was only required for JAWT. Looks to me like it could be dropped for non-arm64 as well.

matthiasblaesing commented on test/com/sun/jna/VarArgsTest.java:
```diff
@@ -42,8 +42,8 @@
-        public void modifyStructureVarArgs(String fmt, Object arg1, Object... args);
-        public String returnStringVarArgs2(String... args);
+        public void modifyStructureVarArgs(String fmt, Object... args);
+        public String returnStringVarArgs2(String fmt, String... args);
```
This needs more explanation - linux arm64 lets you happily call these methods without breaking, so this raises the question what makes mac OS so special here.

fkistner replied:
This is what Apple's documentation says about it:
> The `x86_64` and `arm64` architectures have different calling conventions for variadic functions—functions with a variable number of parameters. On `x86_64`, the compiler treats fixed and variadic parameters the same, placing parameters in registers first and only using the stack when no more registers are available. On `arm64`, the compiler always places variadic parameters on the stack, regardless of whether registers are available. If you implement a function with fixed parameters, but redeclare it with variadic parameters, the mismatch causes unexpected behavior at runtime.

https://developer.apple.com/documentation/apple_silicon/addressing_architectural_differences_in_your_macos_code#overview

matthiasblaesing replied:
Ah - thank you I had to read multiple times to grasp the real change.

fkistner commented:
It is possible to build the Zero variant yourself, if you have access to the hardware: https://github.com/apple/openjdk/

matthiasblaesing commented:
All my questions might have hinted, that I don't own Apple hardware, and I don't see me owning any. For the test error in the callback test, you can try to narrow it down by adding printfs in the testlibrary and/or the native dispatch library to see when it fails or how long it works.

dkocher commented:
I can confirm this is building a universal `libjnidispatch.jnilib` binary on macOS 10.15.5 with Xcode 12.0

```
CC=clang SDKROOT=/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk DEVELOPER_DIR=/Applications/Xcode-beta.app/Contents/Developer/ ant test -Dmake.OPTS="DYNAMIC_LIBFFI=true DARWIN_USE_SDK_LIBFFI=true"
```

matthiasblaesing commented:
So we need a strategy how to proceed:

If I remember correctly, Apple does like fat binaries only very selectively. In principle it should be possible to build a fat binary, that supports x86, x86_64 and arm64. This will need support from XCode to build all three binaries. The alternative would be to build the darwin-x86, darwin-x86-64, darwin-aarch64 separately, and merge only at build time.

We could then also provide a way to drop the darwin special handling (fat binaries). We could bundle the mentioned individual darwin libraries and the fat binary.

What I notice in the invocation above is the usage of the SDK libffi - can we build from the embedded libffi source/can we update it?

tresf commented:
> can we build from the embedded libffi source/can we update it?

@matthiasblaesing this is the strategy I've used in another aarch64 PR (https://github.com/java-native-access/jna/commit/2bf1593fd188dd51f353abfdf98c2a5347cc50cc in #1264).  Note, embedding `libffi`'s project's source code has some snags.  Notably, the `.gitignore` files causes some critical files to not upload (since they're not yet tracked in jna.  Worse, the `.gitignore` is preferred in principal to make tracking source changes easier).  So I've made the assumption in #1264 to include everything except `native/libffi/.git`.  If this is incorrect, guidance is needed.

As an aside, [`submodule`](https://git-scm.com/docs/git-submodule) (or perhaps a [`subtree`](https://www.atlassian.com/git/tutorials/git-subtree) are also viable, but require upstream changes to be accepted first (such as [downstream changes required for `msvcc.sh` to continue working](https://github.com/java-native-access/jna/commit/112d76d4e2feb2b21c36f6753389bb2b88f8acb2)).  I don't want to go too off-topic but wanted to at least mention it incase it's something JNA is interested in.

tresf commented:
> In principle it should be possible to build a fat binary, that supports x86, x86_64 and arm64. 

It appears XCode 10 [dropped 32-bit support](https://stackoverflow.com/a/53961451/3196753), so the following is not possible using XCode 12.  XCode 12 is required for arm64 support.

```bash
clang \
-arch i386 \
-arch x86_64 \
-arch arm64 \
hello.c -o hello
# ld: warning: missing required architecture i386 in file [...]/MacOSX.sdk/usr/lib/libSystem.tbd (3 slices)
# Undefined symbols for architecture i386:
```

XCode 12 can successfully build both `x86_64` and `arm64` if `-arch i386` is removed.

> If I remember correctly, Apple does like fat binaries only very selectively. This will need support from XCode to build all three binaries. The alternative would be to build the darwin-x86, darwin-x86-64, darwin-aarch64 separately, and merge only at build time.

[`lipo`](https://ss64.com/osx/lipo.html) seems to be suited for this task.  A small test shows promise:

```bash
clang -arch x86_64 hello.c -o hello_x86_64
file hello_x86_64
# hello_x86_64 (for architecture x86_64):	Mach-O 64-bit executable x86_64
#                                               ^--- 64-bit Intel.  Good, expected.

clang -arch arm64 hello.c -o hello_aarch64
file hello_aarch64
# hello_aarch64 (for architecture arm64):	Mach-O 64-bit executable arm64
#                                               ^--- 64-bit ARM.  Good, expected.

lipo hello_x86_64 hello_aarch64 -create -output hello
file hello
# hello: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit executable x86_64] [arm64:Mach-O 64-bit executable arm64]
# hello (for architecture x86_64):	Mach-O 64-bit executable x86_64
# hello (for architecture arm64):	Mach-O 64-bit executable arm64
#                                       ^--- Successfully combined.
```

matthiasblaesing commented:
So the option would be: We build the architecture dependent libraries individually on a matching XCode version and update `darwin.jar` instead of replacing it. That way we should get a fat binary, that supports x86, x64 and aarch64. We could then think about also providing a "apple" mode, that drops the fat binary and splits `darwin.jar` into `darwin-x86.jar`, `darwin-x86-64.jar` and `darwin-aarch64.jar`.

In case of an incompatible api change, the whole file will be cleared, rebuilds for mac os need to be build incrementally.

fkistner commented:
> What I notice in the invocation above is the usage of the SDK libffi - can we build from the embedded libffi source/can we update it?

The embedded libffi compiles fine, but the resulting binary does not work correctly for me. Until libffi is updated, using the system libffi seems to be the best option.

> We could then think about also providing a "apple" mode, that drops the fat binary and splits `darwin.jar` into `darwin-x86.jar`, `darwin-x86-64.jar` and `darwin-aarch64.jar`.

Seems to me, the easiest way would to split off a `darwin-aarch64.jar` for now. The decision whether the Intel binary can be separate.

What do you think about this?

fkistner commented:
`darwin-aarch64.jar` can be built successfully on x86_64 using `ant native -Ddynlink.native=true -Dbuild.os.arch=aarch64`.

fkistner commented:
Added `darwin-aarch64.jar` from one of our build agents to give the CI a chance to run. Please advise, if you'd like me to remove it.

matthiasblaesing commented:
> > What I notice in the invocation above is the usage of the SDK libffi - can we build from the embedded libffi source/can we update it?
> 
> The embedded libffi compiles fine, but the resulting binary does not work correctly for me. Until libffi is updated, using the system libffi seems to be the best option.

Please see if rebasing this work onto the work done in https://github.com/java-native-access/jna/pull/1264 helps. That branch holds upstream libffi. I have some hope, that apple did the sane thing and upstreamed its changes. For the system libffi: It is dynamically linked or statically build into the jnidispatch library? What are the licensing terms?

JNA already allows linking against a system libffi, it is the mode, that is used in the debian builds. However, that will utterly fail on non apple platforms (though I don't now if there are any serious darwin users outside apple).

> > We could then think about also providing a "apple" mode, that drops the fat binary and splits `darwin.jar` into `darwin-x86.jar`, `darwin-x86-64.jar` and `darwin-aarch64.jar`.
> 
> Seems to me, the easiest way would to split off a `darwin-aarch64.jar` for now. The decision whether the Intel binary should be split as well can be separate.
> 
> What do you think about this?

I would not do it right now. For me this is either-or, `darwin` was historically the only OS, that was build with fat binaries for JNA. I don't know what triggered that, but it is a fact. A hybrid, where only darwin-aarch64 is split of, would IMHO be an error, that we can not cleanly recover from. We should either split completely or don't. Splitting the darwin binary has also consequences for downstream users and needs to be planned.

> Added `darwin-aarch64.jar` from one of our build agents to give the CI a chance to run. Please advise, if you'd like me to remove it.

In the long run we should see, that the CI infrastructure gets a branch, that is executed on XCode 12, so the binary can be build from there. What I envision is this (only for apple, only while fatbinaries are targetted).

- the build process is first run on XCode 12, then the x64 + aarch64 binaries in the fat jar are updated
- the build process is then run on XCode 9, then the x64 + x86 binaries in the fat jar are updated

We end up with a darwin binary, that supports all three architectures. So could you merge the darwin-aarch64 binary into the darwin binary?

tresf commented:
> the build process is first run on XCode 12, then the x64 + aarch64 binaries in the fat jar are updated

Hi, @vzor- and myself are working on a modification of this PR that does each arch separately, using `unzip|lipo|jar` using `ant` to place only the target architecture back in.  It keeps the fat binary (for now), but takes `lipo` out of the Makefile and places it directly in ant.

I think this is slightly more intuitive than the proposal above which:

* Creates a fat binary with x86_64 + aarch64 using a newer XCode
* Creates a fat binary with x86 + x86_64 using and older XCode
   * First must `lipo` the newer x86_64 back out!

Instead we're focusing on this:

> We build the architecture dependent libraries individually on a matching XCode version and update `darwin.jar` instead of replacing it.

Will share the code soon, probably as a GitHub compare link.

catap commented on build.xml:
```diff
@@ -276,7 +276,13 @@
       <equals arg1="${build.os.family}" arg2="windows"/>
     </condition>
-    <!-- Darwin builds are universal, no arch required -->
+    <condition property="os.prefix" value="darwin-aarch64">
```
A real machine calls arch as `arm64`

tresf replied:
So does Windows, but this project uses `aarch64`.  See also #1264.

catap commented on build.xml:
```diff
@@ -276,7 +276,13 @@
+    <condition property="os.prefix" value="darwin-aarch64">
+      <and>
+        <os family="mac"/>
+        <equals arg1="${jre.arch}" arg2="aarch64"/>
```
And here `arm64`

catap commented on build.xml:
```diff
@@ -564,6 +572,9 @@ osname=macosx;processor=x86;processor=x86-64;processor=ppc
       <zipfileset src="${lib.native}/darwin.jar"
                   includes="*jnidispatch*"
                   prefix="com/sun/jna/darwin"/>
+      <zipfileset src="${lib.native}/darwin-aarch64.jar"
```
I guess and here `arm64`

catap commented on build.xml:
```diff
@@ -912,6 +927,12 @@ osname=macosx;processor=x86;processor=x86-64;processor=ppc
+    <condition property="make.SDKROOT" value="SDKROOT=${xcode}/Developer/SDKs/MacOSX11.0.sdk">
+      <and>
+        <equals arg1="${os.prefix}" arg2="darwin-aarch64" trim="true"/>
+        <available file="${xcode}/Developer/SDKs/MacOSX11.0.sdk"/>
```
12?

tresf commented:
> It is possible to build the Zero variant yourself, if you have access to the hardware: https://github.com/apple/openjdk/

Just an update on this comment, since this was posted, Zulu and Microsoft (yes, Microsoft :smile:)both have Apple Silicon builds available.   At the time of writing this, Microsoft's is broken, so Azul's is [the only precompiled version](https://tinyurl.com/y4n6vxyn).  Microsoft has informed me they'll have a new build out soon.

catap commented:
Let me share a few command from real machine with M1 and macOS 12:

```
catap@Kirills-mini-m1 ~ % uname -m
arm64
catap@Kirills-mini-m1 ~ % uname -a
Darwin Kirills-mini-m1.sa31-cbt.catap.net 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:10 PDT 2020; root:xnu-7195.50.7~2/RELEASE_ARM64_T8101 arm64
catap@Kirills-mini-m1 ~ % 
```

It is called itself `arm64` and not `aarch64` that was a big surprise fr me.

catap commented:
@tresf Microsoft's build doesn't work. It crashed on start. :(

Zulu more stable, but it is `16.ea` and original branch of OpenJDK port doesn't pass all tests.

tresf commented:
> It is called itself arm64 and not aarch64 that was a big surprise fr me.

OpenJDK as well as JNA [call it `aarch64`](http://openjdk.java.net/jeps/391).  Let's not get too tied up with semantics. :)

catap commented:
@tresf indeed.
```
Welcome to Scala 2.13.4 (OpenJDK 64-Bit Server VM, Java 16-ea).
Type in expressions for evaluation. Or try :help.

scala> System.getProperty("os.arch")
val res0: String = aarch64

scala> 
```

tresf commented:
> @tresf Microsoft's build doesn't work. It crashed on start. :(

I've already [posted this information and a link to the details](https://github.com/java-native-access/jna/pull/1238#issuecomment-732333961).

catap commented:
I feel until https://github.com/openjdk/aarch64-port/pull/3 is merged it won't work.

fkistner commented:
> Please see if rebasing this work onto the work done in #1264 helps. That branch holds upstream libffi. I have some hope, that apple did the sane thing and upstreamed its changes.

Confirmed. Building with the libffi from a65d18dcd36b5cd9ee3805257a90ae7132564975 works fine, so no need to link dynamically on macOS. I determined we actually also want https://github.com/libffi/libffi/pull/577 to fix passing of short varargs (see `testShortVarArgs`).

I have the changes ready in [fkistner/jna#darwin_arm64-via-win](https://github.com/fkistner/jna/tree/darwin_arm64-via-win) and will update the PR as soon as https://github.com/java-native-access/jna/pull/1264 is merged.

`DirectCallbacksTest` status is unchanged and still crashes the JRE (might very well be related to https://github.com/java-native-access/jna/pull/1238#issuecomment-732348087), but I don't know enough about Direct Mapping to start digging into it.

fkistner commented:
Added build steps to merge all darwin libs into a single universal fat binary.
Added `darwin-aarch64.jar` (and lipo-ed `darwin.jar`) from build agent.

fkistner commented on build.xml:
```diff
@@ -1111,6 +1107,34 @@ cd ..
     <copy todir="${lib.native}">
       <fileset dir="${build}" includes="${native.jar}"/>
     </copy>
+    <condition property="is_darwin">
+      <matches string="${os.prefix}" pattern="^darwin-"/>
+    </condition>
+    <mkdir if:set="is_darwin" dir="${build}/native-darwin-universal/intel"/>
+    <mkdir if:set="is_darwin" dir="${build}/native-darwin-universal/aarch64"/>
+    <copy if:set="is_darwin" todir="${build}/native-darwin-universal/intel">
+      <zipfileset src="${lib.native}/darwin-intel.jar" includes="*jnidispatch*"/>
+    </copy>
+    <copy if:set="is_darwin" todir="${build}/native-darwin-universal/aarch64">
+      <zipfileset src="${lib.native}/darwin-aarch64.jar" includes="*jnidispatch*"/>
+    </copy>
+    <exec if:set="is_darwin" executable="lipo" dir="${build}/native-darwin-universal" failonerror="true">
+      <arg value="-create"/>
+      <arg value="-output"/>
+      <arg value="libjnidispatch.jnilib"/>
+      <arg value="intel/libjnidispatch.jnilib"/>
+      <arg value="aarch64/libjnidispatch.jnilib"/>
+    </exec>
+    <jar if:set="is_darwin" jarfile="${build}/darwin.jar" createUnicodeExtraFields="never" encoding="UTF-8">
+      <fileset dir="${build}/native-darwin-universal" includes="libjnidispatch.jnilib"/>
+      <manifest>
+        <attribute name="Implementation-Version" value="${jni.version} b${jni.build}"/>
+        <attribute name="Specification-Version" value="${jni.version}"/>
+      </manifest>
+    </jar>
+    <copy if:set="is_darwin" todir="${lib.native}">
+      <fileset dir="${build}" includes="darwin.jar"/>
+    </copy>
```
@matthiasblaesing This might be a bit crude. Please advise, if you'd prefer this e.g. as a `Makefile`, or if you have suggestions how to do this nicer in ant.

fkistner commented:
Tests were broken. Fixed.

fkistner commented on build.xml:
```diff
@@ -1090,6 +1111,35 @@ cd ..
     <copy todir="${lib.native}">
       <fileset dir="${build}" includes="${native.jar}"/>
     </copy>
+    <condition property="is_darwin">
+      <equals arg1="${os.prefix}" arg2="darwin"/>
+    </condition>
+    <property name="build.native.universal" value="${build}/native-darwin-universal"/>
+    <mkdir if:set="is_darwin" dir="${build.native.universal}/intel"/>
+    <mkdir if:set="is_darwin" dir="${build.native.universal}/aarch64"/>
+    <copy if:set="is_darwin" todir="${build.native.universal}/intel">
+      <zipfileset src="${lib.native}/darwin-intel.jar" includes="*jnidispatch*"/>
+    </copy>
+    <copy if:set="is_darwin" todir="${build.native.universal}/aarch64">
+      <zipfileset src="${lib.native}/darwin-aarch64.jar" includes="*jnidispatch*"/>
+    </copy>
+    <exec if:set="is_darwin" executable="lipo" dir="${build.native.universal}" failonerror="true">
+      <arg value="-create"/>
+      <arg value="-output"/>
+      <arg value="libjnidispatch.jnilib"/>
+      <arg value="intel/libjnidispatch.jnilib"/>
+      <arg value="aarch64/libjnidispatch.jnilib"/>
+    </exec>
+    <jar if:set="is_darwin" jarfile="${build}/${os.prefix}.jar" createUnicodeExtraFields="never" encoding="UTF-8">
+      <fileset dir="${build}/native-${os.prefix}-universal" includes="libjnidispatch.jnilib"/>
+      <manifest>
+        <attribute name="Implementation-Version" value="${jni.version} b${jni.build}"/>
+        <attribute name="Specification-Version" value="${jni.version}"/>
+      </manifest>
+    </jar>
```
@matthiasblaesing This might be a bit crude. Please advise, if you'd prefer this e.g. as a `Makefile`, or if you have suggestions how to do this nicer in ant.

tresf commented on CHANGES.md:
```diff
@@ -7,6 +7,7 @@ Next Release (5.7.0)
 
 Features
 --------
+* [#1238](https://github.com/java-native-access/jna/pull/1238): Add macOS `aarch64` architecture to universal `darwin` target. Cherry pick [libffi/libffi#577](https://github.com/libffi/libffi/pull/577). - [@fkistner](https://github.com/fkistner).
```
@fkistner this commit no longer exists in this PR, was that intentional?

fkistner replied:
This PR also requires the bumped libffi from your PR, but for review purposes I decided to only add my changes.
Added the relevant commits to this branch and rebased my changes on top including the cherry-pick.

fkistner commented:
@matthiasblaesing Rebased the PR on @tresf's recent libffi bump and added fresh binaries from build agent as you asked for. We have only good experiences so far with these changes [in production](https://github.com/JetBrains/intellij-deps-jna). However, direct mode still crashes the JVM in some cases and should therefore be considered unavailable for now.
Please advise how to proceed, e.g. if you favor #1297 over these changes.

fkistner commented:
Merged as #1297.
